File: backend/fetch_details.js
const fs = require('fs');
const path = require('path');
const { getStockData } = require('./retriever');

async function fetchAllDetails() {
    console.log("Loading top 1000 list from cache...");
    const inputPath = path.join(__dirname, 'top_1000_companies.json');
    if (!fs.existsSync(inputPath)) {
        console.error("top_1000_companies.json not found. Run scraper first.");
        return;
    }
    
    const companies = JSON.parse(fs.readFileSync(inputPath, 'utf8'));
    console.log(`Starting parallel fetch of details for ${companies.length} stocks...`);
    
    const detailedList = [];
    const limit = 15; // Concurrency limit
    let index = 0;
    
    const startAll = Date.now();
    
    async function worker() {
        while (index < companies.length) {
            const currentIdx = index++;
            const item = companies[currentIdx];
            console.log(`[${currentIdx+1}/${companies.length}] Fetching details for ${item.ticker}...`);
            const start = Date.now();
            try {
                const data = await getStockData(item.ticker);
                
                // Merge rank, country, and CMC market cap
                const merged = {
                    ...data,
                    rank: item.rank,
                    ticker: item.ticker, // Ensure 'ticker' key exists
                    market_cap_formatted: item.market_cap,
                    market_cap_raw: item.market_cap_raw,
                    country: item.country
                };
                detailedList.push(merged);
                console.log(`  Finished ${item.ticker} in ${((Date.now() - start)/1000).toFixed(2)}s`);
            } catch (e) {
                console.error(`  Error fetching ${item.ticker}:`, e.message);
                detailedList.push({
                    rank: item.rank,
                    ticker: item.ticker,
                    company_name: item.company_name,
                    price_per_share: parseFloat(item.share_price.replace(/[$,]/g, '')) || null,
                    market_cap: item.market_cap,
                    market_cap_raw: item.market_cap_raw,
                    country: item.country,
                    error: e.message
                });
            }
            
            // Intermittent save
            if (detailedList.length % 50 === 0) {
                fs.writeFileSync(
                    path.join(__dirname, 'top_1000_companies_detailed.json'),
                    JSON.stringify(detailedList, null, 2)
                );
            }
        }
    }
    
    const promises = [];
    for (let i = 0; i < Math.min(limit, companies.length); i++) {
        promises.push(worker());
    }
    
    await Promise.all(promises);
    
    // Sort by rank
    detailedList.sort((a, b) => (parseInt(a.rank) || 9999) - (parseInt(b.rank) || 9999));
    
    const outputPath = path.join(__dirname, 'top_1000_companies_detailed.json');
    fs.writeFileSync(outputPath, JSON.stringify(detailedList, null, 2));
    
    console.log(`\nCompleted! Fetched ${detailedList.length} stocks in ${((Date.now() - startAll)/1000).toFixed(2)} seconds.`);
    console.log(`Saved database to ${outputPath}`);
    return detailedList;
}

if (require.main === module) {
    fetchAllDetails();
}

module.exports = { fetchAllDetails };

File: fetch_details.py
import sys
import os
import concurrent.futures
import json
import time
import invest_retriever

# Add current dir to path
sys.path.append(os.path.dirname(os.path.abspath(__file__)))

def fetch_stock_details(c_item):
    symbol = c_item["ticker"]
    try:
        data = invest_retriever.get_stock_data(symbol)
        if "error" in data:
            # Fallback to CMC basic data if yfinance fails for a particular stock
            return {
                "rank": c_item.get("rank"),
                "ticker": symbol,
                "company_name": c_item.get("company_name"),
                "price_per_share": c_item.get("share_price"),
                "market_cap": c_item.get("market_cap"),
                "market_cap_raw": c_item.get("market_cap_raw"),
                "country": c_item.get("country"),
                "error": data["error"]
            }
        
        # Merge rank and formatted market cap from CMC
        data["rank"] = c_item.get("rank")
        data["market_cap_formatted"] = c_item.get("market_cap")
        data["market_cap_raw"] = c_item.get("market_cap_raw")
        data["country"] = c_item.get("country")
        return data
    except Exception as e:
        return {
            "rank": c_item.get("rank"),
            "ticker": symbol,
            "company_name": c_item.get("company_name"),
            "price_per_share": c_item.get("share_price"),
            "market_cap": c_item.get("market_cap"),
            "market_cap_raw": c_item.get("market_cap_raw"),
            "country": c_item.get("country"),
            "error": str(e)
        }

def run_fetch():
    print("Loading top 500 list...")
    try:
        with open("top_500_companies.json", "r", encoding="utf-8") as f:
            companies = json.load(f)
    except Exception as e:
        print(f"Error loading list: {e}. Please run get_top_500.py first.")
        return

    print(f"Starting parallel fetch of details for {len(companies)} stocks...")
    start_time = time.time()
    
    detailed_list = []
    
    # Run with 15 parallel workers for high speed
    with concurrent.futures.ThreadPoolExecutor(max_workers=15) as executor:
        futures = {executor.submit(fetch_stock_details, c): c for c in companies}
        
        count = 0
        for future in concurrent.futures.as_completed(futures):
            res = future.result()
            detailed_list.append(res)
            count += 1
            if count % 25 == 0:
                print(f"  Progress: {count}/{len(companies)} stocks loaded...")
                
            # Intermittent autosave in case of interrupts
            if count % 100 == 0:
                with open("top_500_companies_detailed.json", "w", encoding="utf-8") as f:
                    json.dump(detailed_list, f, indent=2)

    # Sort detailed_list by rank so it stays in order
    def get_rank_sort(item):
        r = item.get("rank")
        try:
            return int(r)
        except (ValueError, TypeError):
            return 9999
            
    detailed_list.sort(key=get_rank_sort)

    # Final Save
    output_file = "top_500_companies_detailed.json"
    with open(output_file, "w", encoding="utf-8") as f:
        json.dump(detailed_list, f, indent=2)
        
    print(f"\nCompleted! Fetched {len(detailed_list)} stocks in {time.time() - start_time:.2f} seconds!")
    print(f"Saved to: {output_file}")

if __name__ == "__main__":
    run_fetch()
